tests/run/Tuple-fromProduct.scala

```diff
@@ -0,0 +1,47 @@
+import scala.reflect.ClassTag
+
+object Test {
+  def main(args: Array[String]): Unit = {
+
+    def testProduct(product: Product): Unit = {
+      val t: Tuple = Tuple.fromProduct(product)
+      println(t)
+    }
+
+    testProduct(Tuple1(1))
+    testProduct((1, 2))
+    testProduct((1, 2, 3))
+    testProduct((1, 2, 3, 4))
+    testProduct((1, 2, 3, 4, 5))
+    testProduct((1, 2, 3, 4, 5, 6))
+    testProduct((1, 2, 3, 4, 5, 6, 7))
+    testProduct((1, 2, 3, 4, 5, 6, 7, 8))
+    testProduct((1, 2, 3, 4, 5, 6, 7, 8, 9))
+    testProduct((1, 2, 3, 4, 5, 6, 7, 8, 9, 10))
+    testProduct((1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11))
+    testProduct((1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12))
+    testProduct((1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13))
+    testProduct((1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14))
+    testProduct((1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15))
+    testProduct((1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16))
+    testProduct((1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17))
+    testProduct((1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18))
+    testProduct((1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19))
+    testProduct((1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20))
+    testProduct((1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21))
+    testProduct((1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22))
+    // FIXME ProductN for N > 22 does not extends product
+//    testProduct((1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23))
+//    testProduct((1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24))
+//    testProduct((1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25))
```

Comment:
This seems to be an oversight, I made the following changes, it seems to work:

```Scala
-sealed class *:[+H, +T <: Tuple] extends NonEmptyTuple
+sealed abstract class *:[+H, +T <: Tuple] extends NonEmptyTuple with Product
```

/cc: @odersky

Reply:
I'm looking into it as well. But this might not be what we want. The methods on product should also be specialized. For example `productArity` should do exactly the same as `size`.

Then there is the same issue for `Tuple` in general, not only non empty tuples. There the additional complete cation is that Unit is not a product.

Reply:
I think `Unit` probably _should_ be a `Product` :-)